example/persist.py

```diff
@@ -0,0 +1,190 @@
+        # instead of the supplied initial value, unless the
+        # handler_open_restore flag indicates otherwise.
+        if not handler_open_restore:
+            return
+
+        # We could, in theory, re-apply authentication here if we queried for
+        # that information and then did something with it!
+        res = self._conn.execute("SELECT data FROM pvs WHERE id=?", [self._pv_name])
+        query_val = res.fetchone()
+
+        if query_val is not None:
+            json_val = json.loads(query_val[0])
+            print(f"Will restore to {self._pv_name} value: {json_val['value']}")
+
+            # Override initial value
+            value["value"] = json_val["value"]
+
+            value["timeStamp.secondsPastEpoch"] = json_val["timeStamp"][
+                "secondsPastEpoch"
+            ]
+            value["timeStamp.nanoseconds"] = json_val["timeStamp"]["nanoseconds"]
+        else:
+            # We are using an initial value so persist it
+            self._upsert(value)
+
+    def post(
+        self,
+        pv: SharedPV,
+        value: Value,
+        handler_post_upsert_account=None,
+        handler_post_upsert_peer=None,
+    ):
+        self._update_timestamp(value)
+
+        self._upsert(
+            value,
+            account=handler_post_upsert_account,
+            peer=handler_post_upsert_peer,
+        )
+
+    def put(self, pv: SharedPV, op: ServerOperation):
+        # The post does all the real work, we just add info only available
+        # from the ServerOperation
+        pv.post(
+            op.value(),
+            handler_post_upsert_account=op.account(),
+            handler_post_upsert_peer=op.peer(),
+        )
+        op.done()
+
+    def _update_timestamp(self, value) -> None:
+        if not value.changed("timeStamp") or (
+            value["timeStamp.nanoseconds"] == value["timeStamp.nanoseconds"] == 0
+        ):
+            now = time.time()
+            value["timeStamp.secondsPastEpoch"] = now // 1
+            value["timeStamp.nanoseconds"] = int((now % 1) * 1e9)
+
+    def _upsert(self, value, account=None, peer=None) -> None:
+        # Persist the data; turn into JSON and write it to the DB
+        val_json = json.dumps(value.todict())
+
+        # Use UPSERT: https://sqlite.org/lang_upsert.html
+        conn.execute(
+            """
+            INSERT INTO pvs (id, data, account, peer) 
+                        VALUES (:name, :json_data, :account, :peer)
+            ON CONFLICT(id)
+            DO UPDATE SET data = :json_data, account = :account, peer = :peer;
+            """,
+            {
+                "name": self._pv_name,
+                "json_data": val_json,
+                "account": account,
+                "peer": peer,
+            },
+        )
+        conn.commit()
+
+
+# Create an SQLite dayabase to function as our persistence store
+conn = sqlite3.connect("persist_pvs.db", check_same_thread=False)
+# conn.execute("DROP TABLE IF EXISTS pvs")
```

Comment:
I think this comment is unnecessary, looks like you do the same thing in the next line?